pxd.c

```diff
@@ -859,32 +862,36 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_ext_out *add
 
 	return 0;
 out_disk:
-	put_disk(disk);
+	if (disk) {
+		put_disk(disk);
+	}
 	return err;
 }
 
 static void pxd_free_disk(struct pxd_device *pxd_dev)
 {
 	struct gendisk *disk = pxd_dev->disk;
+	printk(KERN_ALERT "%s: ...", __func__);
 
-	if (!disk)
+	if (!disk) {
 		return;
+	}
 
 	pxd_fastpath_cleanup(pxd_dev);
 	pxd_dev->disk = NULL;
+
+	if (disk) {
+		del_gendisk(disk);
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5,15,0)
-	if (disk_live(disk)) {
-#else	
-	if (disk->flags & GENHD_FL_UP) {
+		blk_cleanup_disk(disk);
```

Comment:
blk_cleanup_disk() already does put_disk(). Likely double free here.
This cleanup path needs to be checked for differing version.

Reply:
Yeah I see that now. Not sure why but it worked on kernel 5.15 when i tried. Anyways updated now.